shell/common/rasterizer.cc

```diff
@@ -497,73 +514,157 @@ Rasterizer::DoDrawResult Rasterizer::DoDraw(
+  // and that view must be the implicit view. Properly support multi-view
+  // in the future.
+  FML_CHECK(tasks.size() == 1u);
+  FML_DCHECK(tasks.front().view_id == kFlutterImplicitViewId);
 
   compositor_context_->ui_time().SetLapTime(
       frame_timings_recorder.GetBuildDuration());
 
-  DlCanvas* embedder_root_canvas = nullptr;
+  // First traverse: Filter out discarded trees
+  auto task_iter = tasks.begin();
+  while (task_iter != tasks.end()) {
+    if (delegate_.ShouldDiscardLayerTree(task_iter->view_id,
+                                         *task_iter->layer_tree)) {
+      last_draw_statuses_[task_iter->view_id] = DrawSurfaceStatus::kDiscarded;
+      task_iter = tasks.erase(task_iter);
+    } else {
+      ++task_iter;
+    }
+  }
+  if (tasks.empty()) {
+    frame_timings_recorder.RecordRasterStart(fml::TimePoint::Now());
+    frame_timings_recorder.RecordRasterEnd();
+    return nullptr;
+  }
+
   if (external_view_embedder_) {
     FML_DCHECK(!external_view_embedder_->GetUsedThisFrame());
     external_view_embedder_->SetUsedThisFrame(true);
     external_view_embedder_->BeginFrame(
-        layer_tree.frame_size(), surface_->GetContext(), device_pixel_ratio,
-        raster_thread_merger_);
-    embedder_root_canvas = external_view_embedder_->GetRootCanvas();
+        // TODO(dkwingsmt): Add all views here.
+        tasks.front().layer_tree->frame_size(), surface_->GetContext(),
+        tasks.front().device_pixel_ratio, raster_thread_merger_);
+  }
+
+  std::optional<fml::TimePoint> presentation_time = std::nullopt;
+  // TODO (https://github.com/flutter/flutter/issues/105596): this can be in
+  // the past and might need to get snapped to future as this frame could
+  // have been resubmitted. `presentation_time` on SubmitInfo is not set
+  // in this case.
+  {
+    const auto vsync_target_time = frame_timings_recorder.GetVsyncTargetTime();
+    if (vsync_target_time > fml::TimePoint::Now()) {
+      presentation_time = vsync_target_time;
+    }
   }
 
   frame_timings_recorder.RecordRasterStart(fml::TimePoint::Now());
 
+  // Second traverse: draw all layer trees.
+  std::list<LayerTreeTask> resubmitted_tasks;
+  for (LayerTreeTask& task : tasks) {
+    int64_t view_id = task.view_id;
+    std::unique_ptr<LayerTree> layer_tree = std::move(task.layer_tree);
+    float device_pixel_ratio = task.device_pixel_ratio;
+
+    DrawSurfaceStatus status = DrawToSurfaceUnsafe(
+        view_id, *layer_tree, device_pixel_ratio, presentation_time);
+
+    last_draw_statuses_[view_id] = status;
+    if (status == DrawSurfaceStatus::kSuccess) {
+      last_successful_tasks_.insert_or_assign(
+          view_id,
+          LayerTreeTask(view_id, std::move(layer_tree), device_pixel_ratio));
+    } else if (status == DrawSurfaceStatus::kRetry) {
+      resubmitted_tasks.emplace_back(view_id, std::move(layer_tree),
+                                     device_pixel_ratio);
+    }
+  }
+  // TODO(dkwingsmt): Pass in all raster caches
```

Comment:
File a bug.

What's not clear to me is whether we'd actually want multiple raster caches or just one raster cache per layer tree. There are also ongoing explorations to see if we can just remove the raster cache entirely as we migrate to Impeller.

Reply:
(@jonahwilliams fyi)

Reply:
I don't think its possible for the rasterizer to know if it is safe to share any of this information. i.e. if these are from different windows with a different GlContext, or the different windows with the same GlContext, et cetera. The backends are not designed to handle this.